File: day21/.js
function clearGifts(warehouse, drops) {
  const rows = warehouse.length
  const cols = warehouse[0].length
  let newWarehouse = warehouse.map(row => [...row])

  for (const col of drops) {
    let placedRow = -1
    for (let row = rows - 1; row >= 0; row--) {
      if (newWarehouse[row][col] === ".") {
        newWarehouse[row][col] = "#"
        placedRow = row
        break
      }
    }

    if (placedRow != -1 && newWarehouse[placedRow].every(cell => cell === "#")) {
      newWarehouse.splice(placedRow, 1)
      newWarehouse.unshift(new Array(cols).fill("."))
    }
  }

  return newWarehouse
}

File: day21/.py
def clear_gifts(warehouse: list[list[str]], drops: list[int]) -> list[list[str]]:
  rows = len(warehouse)
  cols = len(warehouse[0])
  newWarehouse = [row[:] for row in warehouse]

  for col in drops:
    placedRow = -1
    for row in range(rows-1, -1, -1):
      if newWarehouse[row][col] == ".":
        newWarehouse[row][col] = "#"
        placedRow = row
        break

    if placedRow != -1 and all(cell == "#" for cell in newWarehouse[placedRow]):  
      newWarehouse.pop(placedRow)
      newWarehouse.insert(0, ["." for _ in range(cols)])

  return newWarehouse
